### DISCUSSION TITLE
Where is the `wasmOffsetData` come from?

### DISCUSSION THREAD

**Discussion by junyuecao** [emscripten-core/emscripten#19985] [Q&A]
I was trying to run asan builds with pthreads and custom instantiateWasm. 
I don't see any demo or test case that shows how to construct this data.

https://github.com/emscripten-core/emscripten/blob/0538c9dba690359baa7c09e99a7ff597796c4939/src/preamble.js#L1092-L1101

I only see a demo like this, so is this the way to inject a WasmOffsetConverter?
https://github.com/emscripten-core/emscripten/blob/0538c9dba690359baa7c09e99a7ff597796c4939/test/manual_wasm_instantiate.html#L178

**Comment by sbc100** [comment]
Is the problem you are seeing the the assert on line 1097 of `preamble.js` is firing?   i.e. is it 
`Module['wasmOffsetData']` that is missing?

**Comment by junyuecao** [comment]
> Is the problem you are seeing the the assert on line 1097 of `preamble.js` is firing? i.e. is it
> `Module['wasmOffsetData']` that is missing?

@sbc100  Do you mean that I should construct  a new WasmOffsetConverter and set to the `Module['wasmOffsetData']` ? 
The problem is constructing a WasmOffsetConverter need the `output.module` which is not ready before calling `Module['instantiateWasm']` (ling 1113)
https://github.com/emscripten-core/emscripten/blob/0538c9dba690359baa7c09e99a7ff597796c4939/src/preamble.js#L1099-L1113

I was wondering if we should move constructing WasmOffsetConverter to 
https://github.com/emscripten-core/emscripten/blob/0538c9dba690359baa7c09e99a7ff597796c4939/src/preamble.js#L951

**Reply by sbc100** [reply]
Sorry no, I wasn't suggesting anything, I was asking a specific questions in order to better understand what is going on: `Is the problem you are seeing that the assert on line 1097 of preamble.js is firing? `

**Comment by junyuecao** [comment]
@sbc100 

> Is the problem you are seeing that the assert on line 1097 of preamble.js is firing?

It's not about the assert on line 1097. Let me try to explain the problem:
If I didn't set `wasmOffsetData` , here it is undefined
![image](https://github.com/emscripten-core/emscripten/assets/1000408/2927c368-ae6f-44f0-b167-12356a1e6721)

then `resetPrototype` will return an object without offset data so that asan can not get the func name of an offset.
![image](https://github.com/emscripten-core/emscripten/assets/1000408/8581153d-c0ea-4b20-ad78-ce4afca9af9b)

So it means I should provide an `Module["wasmOffsetData"]` for `resetPrototype` to retrurn a correct `WasmOffsetConverter`.
**Here is the problem**: the constructor of `WasmOffsetConverter` accept wasmBytes and **wasmModule**. But I can't get the **wasmModule** before `Module["instantiateWasm"]` is called. 

```
function WasmOffsetConverter(wasmBytes, wasmModule) {}
```

I think it may work to put it in the `function receiveInstance(instance, module)` so that I have a chance to constructing an `WasmOffsetConverter` in my custom `Module["instantiateWasm"]`.
